File: 2296.design-a-text-editor.cpp
#\n# @lc app=leetcode id=2296 lang=cpp\n#\n# [2296] Design a Text Editor\n#\n# @lc code=start\nclass TextEditor {\npublic:\n    std::string leftText;\n    std::string rightText;\n    \n    TextEditor() {}\n    \n    void addText(std::string text) {\n        leftText += text;\n    }\n    \n    int deleteText(int k) {\n        int count = std::min(k, static_cast<int>(leftText.size()));\n        leftText.erase(leftText.end() - count, leftText.end());\n        return count;\n    }\n    \n    std::string cursorLeft(int k) {\n        int moveCount = std::min(k, static_cast<int>(leftText.size()));\n        rightText.insert(rightText.begin(), leftText.end() - moveCount, leftText.end());\n        leftText.erase(leftText.end() - moveCount, leftText.end());\n        return getLeftContext();   \nn}\nnstd::string cursorRight(int k) {nnint moveCount = std::min(k, static_cast<int>(rightText.size()));nnleftText.append(rightText.begin(), rightText.begin() + moveCount);nnrightText.erase(rightText.begin(), rightText.begin() + moveCount);nnreturn getLeftContext();   \nn}nnprivate:nnstd::string getLeftContext() const {nnint len = std::min(10, static_cast<int>(leftText.size()));nnreturn leftText.substr(leftText.size() - len); nn}\nn};

/**
 * Your TextEditor object will be instantiated and called as such:
 * TextEditor* obj = new TextEditor();
 * obj->addText(text);
 * int param_2 = obj->deleteText(k);
 * std::string param_3 = obj->cursorLeft(k);
 * std::string param_4 = obj->cursorRight(k);
 */
# @lc code=end

File: 2296.design-a-text-editor.java
#
# @lc app=leetcode id=2296 lang=java
#
# [2296] Design a Text Editor
#

# @lc code=start
import java.util.*;
class TextEditor {
    Deque<Character> left;
    Deque<Character> right;
    public TextEditor() {
        left = new ArrayDeque<>();
        right = new ArrayDeque<>();
    }
    public void addText(String text) {
        for (char c : text.toCharArray()) {
            left.addLast(c);
        }
    }
    public int deleteText(int k) {
        int deleted = 0;
        while (deleted < k && !left.isEmpty()) {
            left.removeLast();
            deleted++;
        }
        return deleted;
    }
    public String cursorLeft(int k) {
        int moved = 0;
        while (moved < k && !left.isEmpty()) {
            right.addFirst(left.removeLast());
            moved++;
        }
        return getLast10();
    }
    public String cursorRight(int k) {
        int moved = 0;
        while (moved < k && !right.isEmpty()) {
            left.addLast(right.removeFirst());
            moved++;
        }
        return getLast10();
    }
    private String getLast10() {
        StringBuilder sb = new StringBuilder();
        Iterator<Character> it = left.descendingIterator();
        int cnt = 0;
        while (it.hasNext() && cnt < 10) {
            sb.append(it.next());
            cnt++;
        }
        return sb.reverse().toString();
    }
}
/**
* Your TextEditor object will be instantiated and called as such:
* TextEditor obj = new TextEditor();
* obj.addText(text);
* int param_2 = obj.deleteText(k);
* String param_3 = obj.cursorLeft(k);
* String param_4 = obj.cursorRight(k);
*/
# @lc code=end